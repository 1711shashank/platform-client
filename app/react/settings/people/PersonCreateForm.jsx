import React from "react";
import { Link, Redirect, BrowserRouter } from "react-router-dom";
import PropTypes from "prop-types";
import InlineLoading from "react/common/ui/InlineLoading";
// import Router from "react-router";

const propTypes = {
    saveNewPerson: PropTypes.func.isRequired,
    roles: PropTypes.arrayOf(PropTypes.object).isRequired,
    isLoadingRoles: PropTypes.bool.isRequired
    // error: PropTypes.shape({
    //     message: PropTypes.string
    // }).isRequired
};

class PersonCreateForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            form: {
                email: "howcomethisgoesaway",
                password: "",
                realname: "",
                role: { display_name: "Select a role", name: "" }
            },
            redirectToPeopleList: false
        };
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.handleOptionChange = this.handleOptionChange.bind(this);
        this.renderRoles = this.renderRoles.bind(this);
    }

    handleChange(event) {
        this.setState({
            form: {
                ...this.state.form,
                [event.target.id]: event.target.value
            }
        });
    }

    handleOptionChange(event) {
        this.setState({
            form: {
                ...this.state.form,
                [event.target.id]: {
                    display_name: event.target.value,
                    name: event.target.id
                }
            }
        });
    }

    handleSubmit(e) {
        e.preventDefault();
        const person = Object.assign({}, this.state.form);

        // TODO: do this in a cool ES6 way
        person.role = this.state.form.role.name;
        this.props.saveNewPerson(person).then(() => {
            this.setState(() => ({
                redirectToPeopleList: true,
                form: {}
            }));
        });
        // after success, clear state
        // after failure, state remains
        // TBD after we figure out endpoints
    }

    renderRoles() {
        return (
            <label htmlFor="role">
                Role
                <select
                    id="role"
                    value={this.state.form.role.display_name}
                    onChange={this.handleOptionChange}
                >
                    <option
                        id="default"
                        value={this.state.form.role.display_name}
                    >
                        {this.state.form.role.display_name}
                    </option>
                    {this.props.roles
                        .filter(
                            item =>
                                item.display_name !==
                                this.state.form.role.display_name
                        )
                        .map(role => (
                            <option
                                key={role.id}
                                id={role.name}
                                value={role.display_name}
                            >
                                {role.display_name}
                            </option>
                        ))}
                </select>
            </label>
        );
    }

    render() {
        if (this.state.redirectToPeopleList === true) {
            return <Redirect to="/settings/users" />;
        }

        return (
            <BrowserRouter>
                <main role="main">
                    <div>
                        <h3>Add people to Ushahidi</h3>
                        <p>
                            Add members of your team, stakeholders, and other
                            members of your community to Ushahidi.
                        </p>
                    </div>
                    <form onSubmit={this.handleSubmit}>
                        <label htmlFor="realname">
                            Name
                            <input
                                type="text"
                                placeholder="What is this person's full name"
                                id="realname"
                                value={this.state.form.realname}
                                onChange={this.handleChange}
                                required
                            />
                        </label>
                        <label htmlFor="email">
                            Email
                            <input
                                type="text"
                                placeholder="email"
                                id="email"
                                value={this.state.form.email}
                                onChange={this.handleChange}
                                required
                            />
                        </label>
                        <label htmlFor="password">
                            Password
                            <input
                                type="password"
                                placeholder="password"
                                id="password"
                                value={this.state.form.password}
                                onChange={this.handleChange}
                                required
                            />
                        </label>
                        {this.props.isLoadingRoles ? (
                            <InlineLoading />
                        ) : (
                            this.renderRoles()
                        )}
                        <Link to="/settings/users">
                            <button className="button-beta">Cancel</button>
                        </Link>
                        <button type="submit">Submit</button>
                    </form>
                </main>
            </BrowserRouter>
        );
    }
}

PersonCreateForm.propTypes = propTypes;

export default PersonCreateForm;
