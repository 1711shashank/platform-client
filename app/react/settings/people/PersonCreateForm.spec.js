import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import PersonCreateForm from "./PersonCreateForm";

test("a Person Form with roles displays roles correctly", () => {
    const saveNewPerson = () => {};
    const roles = [{ id: "email", name: "email" }];
    const isLoading = {REQUEST_ROLES: false};
    const component = renderer.create(
        <PersonCreateForm
            saveNewPerson={saveNewPerson}
            roles={roles}
            isLoading={isLoading}
        />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("A Person Form without roles displays loading instead", () => {
    const saveNewPerson = () => {};
    const roles = [];
    const isLoadingRoles = true;
    const component = renderer.create(
        <PersonCreateForm
            saveNewPerson={saveNewPerson}
            roles={roles}
            isLoadingRoles={isLoadingRoles}
        />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("the handleChange method updates the input value", () => {
    const saveNewPerson = () => {};
    const roles = [{ id: "email", name: "email" }];
    const isLoadingRoles = false;

    const personCreateForm = shallow(
        <PersonCreateForm
            saveNewPerson={saveNewPerson}
            roles={roles}
            isLoadingRoles={isLoadingRoles}
        />
    );

    expect(personCreateForm.find('[id="realname"]').props().value).toBe("");
    const realname = personCreateForm.find('[id="realname"]');
    realname.simulate("change", {
        target: { value: "My real name", id: "realname" }
    });
    expect(personCreateForm.find('[id="realname"]').props().value).toBe(
        "My real name"
    );
});

// test for submit success

// test for submit error
