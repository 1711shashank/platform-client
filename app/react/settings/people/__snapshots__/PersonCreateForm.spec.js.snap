// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Person Form without roles displays loading instead 1`] = `
<main
  role="main"
>
  <div>
    <h3>
      Add people to Ushahidi
    </h3>
    <p>
      Add members of your team, stakeholders, and other members of your community to Ushahidi.
    </p>
  </div>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="realname"
    >
      Name
      <input
        id="realname"
        onChange={[Function]}
        placeholder="What is this person's full name"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="email"
    >
      Email
      <input
        id="email"
        onChange={[Function]}
        placeholder="email"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        id="password"
        onChange={[Function]}
        placeholder="password"
        required={true}
        type="password"
        value=""
      />
    </label>
    <p>
      Loading...
    </p>
    <a
      href="/settings/users"
      onClick={[Function]}
    >
      <button
        className="button-beta"
      >
        Cancel
      </button>
    </a>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</main>
`;

exports[`a Person Form with roles displays roles correctly 1`] = `
<main
  role="main"
>
  <div>
    <h3>
      Add people to Ushahidi
    </h3>
    <p>
      Add members of your team, stakeholders, and other members of your community to Ushahidi.
    </p>
  </div>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="realname"
    >
      Name
      <input
        id="realname"
        onChange={[Function]}
        placeholder="What is this person's full name"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="email"
    >
      Email
      <input
        id="email"
        onChange={[Function]}
        placeholder="email"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        id="password"
        onChange={[Function]}
        placeholder="password"
        required={true}
        type="password"
        value=""
      />
    </label>
    <label
      htmlFor="role"
    >
      Role
      <select
        id="role"
        onChange={[Function]}
        value=""
      >
        <option
          default={true}
          disabled={true}
          value=""
        >
          Select A Value...
        </option>
        <option
          selected=""
          value="email"
        />
      </select>
    </label>
    <a
      href="/settings/users"
      onClick={[Function]}
    >
      <button
        className="button-beta"
      >
        Cancel
      </button>
    </a>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</main>
`;
